### PR TITLE
Layers 16 - Tests for Update and Delete Layer API

### LINKED ISSUE #1827: Test Scenarios Update and Delete API Layer
# Test Scenarios for UPDATE and DELETE Layer APIs
## UPDATE LAYER API Test Scenarios
A. Positive Scenarios (8 tests)

- test_update_layer_change_tunables - Update layer with modified tunables (add/remove/modify)
- test_update_layer_change_bounds - Update tunable bounds (upper_bound, lower_bound, step)
- test_update_layer_change_presence_type - Change from presence='always' to queries
- test_update_layer_add_queries - Add new queries to existing layer
- test_update_layer_modify_queries - Modify existing query strings
- test_update_layer_change_details - Update layer details/description
- test_update_layer_add_tunable - Add new tunable to existing layer
- test_update_layer_remove_tunable - Remove tunable from layer

B. Negative Scenarios (12 tests)

- test_update_layer_non_existent - Update layer that doesn't exist (404)
- test_update_layer_empty_name - Update with empty layer_name
- test_update_layer_null_tunables - Update with null tunables array
- test_update_layer_invalid_tunable_bounds - Update with invalid bounds
- test_update_layer_duplicate_tunable_names - Update creating duplicate tunable names
- test_update_layer_remove_all_tunables - Update removing all tunables (should fail)
- test_update_layer_invalid_presence_config - Update with invalid layer_presence
- test_update_layer_name_mismatch - URL name != body name
- test_update_layer_missing_required_fields - Update missing mandatory fields
- test_update_layer_invalid_query_syntax - Update with invalid PromQL
- test_update_layer_change_to_invalid_value_type - Change tunable value_type incorrectly
- test_update_layer_with_experiments_active - Update layer currently used in active experiments

C. Validation Scenarios (6 tests)

- test_update_layer_preserves_metadata - Verify metadata is preserved
- test_update_layer_version_increment - Check if version is incremented: Skipped for now
- test_update_layer_audit_trail - Verify update is logged/audited : Requires infrastructure (logs/audit tables). Not suitable for API functional tests
- test_update_layer_partial_update - Update only specific fields
- test_update_layer_idempotency - Same update twice should be idempotent
- test_update_layer_concurrent_updates - Handle concurrent update requests : Not being implemented now due Complexity in implementing. Better as performance/stress test 

D. Edge Cases (4 tests)

- test_update_layer_very_large_payload - Update with large number of tunables
- test_update_layer_special_characters - Update with special chars in strings
- test_update_layer_unicode_content - Update with unicode characters

DELETE LAYER API Test Scenarios
A. Positive Scenarios (5 tests)

- test_delete_layer_success - Delete existing layer successfully
- test_delete_layer_verify_removal - Verify layer is removed from list, and also verifies Delete multiple layers sequentially
- test_delete_layer_recreate - Delete and recreate layer with same name
- test_delete_layer_all_types - Delete layers with different presence types

B. Negative Scenarios (8 tests)

- test_delete_layer_non_existent - Delete layer that doesn't exist (404)
- test_delete_layer_empty_name - Delete with empty layer name
- test_delete_layer_null_name - Delete with null layer name
- test_delete_layer_invalid_name - Delete with invalid characters in name
- test_delete_layer_missing_name_param - Delete without name parameter
- test_delete_layer_with_active_experiments - Delete layer used in active experiments (should fail)
- test_delete_layer_twice - Delete same layer twice (second should fail)
- test_delete_layer_sql_injection - SQL injection attempt in layer name

C. Validation Scenarios (5 tests)

- test_delete_layer_cascading_effects - Verify experiments are updated after deletion  - covered by delete layer with active experiment
- test_delete_layer_audit_trail - Verify deletion is logged - Not Suitable
- test_delete_layer_permissions - Test deletion with different permission levels - Needs auth 
- test_delete_layer_soft_vs_hard_delete - Verify if deletion is soft or hard - covered by delete_layer_recreate
- test_delete_layer_rollback - Test rollback if deletion fails mid-process -- Complex scenario skipping

D. Integration Scenarios (4 tests)

- test_delete_layer_experiment_cleanup - Verify experiments handle deleted layers- Already covered by del_layer_active_Experiment
- test_delete_layer_detection_impact - Verify detection stops after deletion - complex for use case
- test_delete_layer_recommendations_impact - Check impact on existing recommendations -- integration tests not suitable for current suite
- test_delete_layer_concurrent_delete - Handle concurrent delete requests -- better as stress test
 
E. Edge Cases (3 tests)

- test_delete_layer_case_sensitivity - Test case-sensitive layer name deletion
- test_delete_layer_with_special_chars - Delete layer with special characters in name
- test_delete_layer_performance - Delete performance with many layers - skipped for functional test suite

### PR DESCRIPTION
**Pull request by bhanvimenghani**
## Description

This Test PR adds positive negative and other edge case tests for Update and Delete Layers API. More details on each tests that has been added can be found in the issue #1827 

Fixes # (issue)

### Type of change

- [ ] Bug fix
- [ ] New feature
- [ ] Docs update
- [ ] Breaking change (What changes might users need to make in their application due to this PR?)
- [ ] Requires DB changes

## How has this been tested?

Please describe the tests that were run to verify your changes and steps to reproduce. Please specify any test configuration required. 

- [ ] New Test X
- [ ] Functional testsuite

**Test Configuration**
* Kubernetes clusters tested on: 

## Checklist :dart:

- [ ] Followed coding guidelines
- [ ] Comments added
- [ ] Dependent changes merged
- [ ] Documentation updated
- [ ] Tests added or updated

## Additional information

Include any additional information such as links, test results, screenshots here

## Summary by Sourcery

Add full CRUD support and enhanced validation for layer management APIs, along with comprehensive tests for create, list, update, and delete operations.

New Features:
- Introduce updateLayer and deleteLayer REST endpoints for managing existing layers.
- Add helper functions for updating, deleting, and bulk-cleaning layers via the test helpers.

Bug Fixes:
- Ensure createLayer rejects empty request bodies and surfaces validation errors with consistent JSON error responses.
- Align tunable validation error messages and handling with the new validation logic, including bounds, step, and type errors.

Enhancements:
- Standardize success and error responses for layer APIs to use structured JSON and appropriate HTTP status codes.
- Tighten input validation for layer creation and updates, including required fields, apiVersion/kind checks, presence configuration, queries, labels, and tunables.
- Extend DAO and metrics support to cover layer update and delete operations.

Tests:
- Expand createLayer tests with extensive positive and negative coverage for optional fields, queries, labels, tunables, and API structure.
- Refactor listLayers tests to use the new delete and cleanup helpers and add performance, case-sensitivity, and sorting scenarios.
- Add comprehensive updateLayer test suite covering successful updates, mandatory fields, invalid configurations, and edge cases.
- Add comprehensive deleteLayer test suite covering successful deletions, error scenarios, and safety against malformed input.

### CODE REVIEW

File: src/main/java/com/autotune/database/dao/ExperimentDAOImpl.java [hunk 1]
```diff
@@ -689,6 +689,83 @@ public ValidationOutputData addLayerToDB(KruizeLMLayerEntry kruizeLayerEntry) {
         return validationOutputData;
```
[the remark this answers is not in the document]
Comment (bhanvimenghani) [reply src/main/java/com/autotune/database/dao/ExperimentDAOImpl.java:689]: not related to changes in this pr, will be accommodated in previous dev changes prs

**Comment by kusumachalasani** [comment]
@bhanvimenghani Can you please submit the tests in jenkins and update the results here.

**Comment by chandrams** [comment]
@bhanvimenghani  Are you validating if the update layer API updates are actually getting used?

File: tests/scripts/local_monitoring_tests/rest_apis/test_update_layer.py [hunk 2]
```diff
@@ -0,0 +1,1214 @@
...
+    tmp_create = f"/tmp/create_{layer_name}.json"
+    with open(tmp_create, "w") as f:
+        json.dump(create_data, f)
+
+    create_response = create_layer(tmp_create)
+    assert create_response.status_code == SUCCESS_STATUS_CODE
+
+    # Update layer - change bounds
+    update_data = copy.deepcopy(create_data)
+    update_data['tunables'][0]['upper_bound'] = "4096"
```
Comment (chandrams) [inline tests/scripts/local_monitoring_tests/rest_apis/test_update_layer.py:105]: @bhanvimenghani There is no validation of the updated values set.

Comment (bhanvimenghani) [reply tests/scripts/local_monitoring_tests/rest_apis/test_update_layer.py:105]: Thanks for pointing out, I have added a validation check to check for updations.

**Comment by chandrams** [comment]
@bhanvimenghani Can you please share the test run links on jenkins

**Comment by chandrams** [comment]
> @bhanvimenghani Can you please share the test run links on jenkins

Can you share the test results.

@shreyabiradar07 Can you please review this as well

File: tests/scripts/helpers/kruize.py [hunk 3]
```diff
@@ -738,63 +738,84 @@ def list_layers(layer_name=None, logging=True):
     return response
 
 
-# Description: This function deletes a layer from the database (temporary workaround until deleteLayer API is implemented)
+# Description: This function deletes a layer using DELETE /deleteLayer API
```
Comment (shreyabiradar07) [inline tests/scripts/helpers/kruize.py:741]: Update the function descriptions based on the API functionality

**Review by shreyabiradar07 — CHANGES_REQUESTED** [review]
